src/Mod/Sketcher/App/Sketch.cpp

```diff
@@ -2435,6 +2454,35 @@ int Sketch::addTangentConstraint(int geoId1, int geoId2)
     return -1;
 }
 
+int Sketch::addTangentLineAtBSplineKnotConstraint(int checkedlinegeoId, int checkedbsplinegeoId, int checkedknotgeoid)
+{
+    GCS::BSpline &b = BSplines[Geoms[checkedbsplinegeoId].index];
+    GCS::Line &l = Lines[Geoms[checkedlinegeoId].index];
+
+    size_t knotindex = b.knots.size();
+
+    auto knotIt = std::find(b.knotpointGeoids.begin(),
+                    b.knotpointGeoids.end(), checkedknotgeoid);
+
+    knotindex = std::distance(b.knotpointGeoids.begin(), knotIt);
+
+    if (knotindex >= b.knots.size()){
+        Base::Console().Error("addConstraint: Knot index out-of-range!\n");
+        return -1;
+    }
+
+    if (b.mult[knotindex] >= b.degree && (b.periodic || (knotindex > 0 && knotindex < (b.knots.size()-1)))) {

```

Comment:
@abdullahtahiriyo if the first is true but the second isn't, that means we have first or last knot. When this code was part of `addAngleAtPointConstraint`, they would pass through to the "classical" method, which uses the `AngleAtPoint` solver constraint. Is it fine if we call that method from here to handle those corner cases?

Notes:
1. Typically these knots are behind the start/end vertices of the curve, so these knots have to be selected specifically through the taskview or in console.
2. The way you have defined it here, "checked" geoId's are passed to this method. We also need to consider if that's fine to do. I believe that only requires `checkGeoId(checkGeoId(x)) == checkGeoId(x)`.